pkg/cluster/edit/diff.go

```diff
@@ -29,3 +38,62 @@ func ShowDiff(t1 string, t2 string, w io.Writer) {
 
 	fmt.Fprint(w, dmp.DiffPrettyText(diffs))
 }
+
+// ValidateSpecDiff checks and validates the new spec to see if the modified
+// keys are all marked as editable
+func ValidateSpecDiff(s1, s2 interface{}) error {
+	differ, err := diff.NewDiffer(
+		diff.TagName(validateTagName),
+	)
+	if err != nil {
+		return err
+	}
+	changelog, err := differ.Diff(s1, s2)
+	if err != nil {
+		return err
+	}
+
+	if len(changelog) == 0 {
+		return nil
+	}
+
+	msg := make([]string, 0)
+	for _, c := range changelog {
+		if len(c.Path) > 0 {
+			// c.Path will be the tag value if TagName matched on the field
+			if c.Type == diff.UPDATE && c.Path[len(c.Path)-1] == validateTagEditable {
+				// If the field is marked as editable, it is allowed to be modified no matter
+				// its parent level element is marked as editable or not
+				continue
+			}
+
+			pathEditable := true
+			pathIgnore := false
+			for _, p := range c.Path {
+				if _, err := strconv.Atoi(p); err == nil {
```

Comment:
I think we cannot ignore the slice, for example, user add a new host into `tidb_servers`.

Reply:
`tidb_servers` is not ignored, `server_configs.tidb` is.